### PR TITLE
Arrow functionality support - Addressed review comments and added few Integration test cases

### PR DESCRIPTION
**Pull request by Sridhar-Durgam2022**
Addressed review comments.
Added Integration test cases

### CODE REVIEW

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		config.setBigQueryReadTable("wordcount_output");
+		config.setProjectId("q-gcp-6750-pso-gs-flink-22-01");
+		config.setDataset("wordcount_dataset");
+		String srcQueryString = "CREATE TABLE " + config.getBigQueryReadTable() + " (word STRING , word_count BIGINT)";
+		flinkTableEnv.executeSql(srcQueryString + "\n" + "WITH (\n" + "  'connector' = 'bigquery',\n"
+				+ "  'format' = 'arrow',\n" + "  'configOptions' = '" + config.getConfigMap() + "'\n" + ")");
+		Table result = flinkTableEnv.from(config.getBigQueryReadTable());
+		Table datatable = result.select($("word"), $("word_count"));
+		TableResult tableapi = datatable.execute();		
+		assertNotNull(tableapi);
```
Comment (davidrabinowitz) [inline src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java:84]: Please convert to google-truth asserts as they are more readable:
```
assertThat(tableapi).isNotNull();
assertThat(tableapi.getTableSchema().getFieldCount()).isEqualTo(2);
```

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
* The common sub package should be deleted, those classes are in the `bigquery-connector-common` library
* spaces, not tabs. Please keep the original indentation
* Flink should be at version 1.11.0
* There should be no specific project references in the code
* Run `./mvnw spotless:apply` on the code
* When reading from query, the options `viewsEnabled` and `materializationDataset` are mandatory

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	public FlinkReadIntegrationTest() {
+		super();
+		config.setGcpCredentialKeyFile(System.getenv("GOOGLE_APPLICATION_CREDENTIALS"));
+		config.setProjectId("q-gcp-6750-pso-gs-flink-22-01");
+		config.setDataset("wordcount_dataset");
+		config.setBigQueryReadTable("wordcount_output");
+	}
+
+	private void testWordCount(TableResult tableRes) {
```
Comment (davidrabinowitz) [inline src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java:36]: s/tableRes/tableResult

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		config_datatype.setDataset("test");
+		config_datatype.setBigQueryReadTable("data_types_test");
+		config_datatype.setSelectedFields("string_datatype,bytes_datatype,integer_datatype,float_datatype,boolean_datatype");
+		String srcQueryString = "CREATE TABLE " + config_datatype.getBigQueryReadTable()
+				+ " (string_datatype STRING , bytes_datatype BYTES, integer_datatype INTEGER, float_datatype FLOAT, boolean_datatype BOOLEAN)"; // numeric_datatype NUMERIC
+		flinkTableEnv.executeSql(srcQueryString + "\n" + "WITH (\n" + "  'connector' = 'bigquery',\n"
+				+ "  'format' = 'arrow',\n" + "  'configOptions' = '" + config_datatype.getConfigMap() + "'\n" + ")");
+		Table result = flinkTableEnv.from(config_datatype.getBigQueryReadTable());
+		TableResult tableapi = result.execute();
+		tableapi.print();
```
Comment (davidrabinowitz) [inline src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java:58]: remove print()

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	private void testReadFilter(TableResult tableRes) {
+		assertThat(tableRes).isEqualTo("Empty Set");
+	}
+
+	@Test
+	public void testReadForDifferentDataTypes() {
+
+		Configuration config_datatype = new Configuration();
```
Comment (davidrabinowitz) [inline src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java:47]: please use java naming style

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.flink.bigquery.model.Configuration;
+
+public class FlinkBigQueryIntegrationTestBase {
+
+	private BigQuery bq;
+	public static StreamTableEnvironment flinkTableEnv;
+	public static Configuration config = new Configuration();
```
Comment (davidrabinowitz) [inline src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java:29]: why to have the configuration field if each test defines one of their own?

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		this.bq = BigQueryOptions.getDefaultInstance().getService();
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.setParallelism(1); // source only supports parallelism of 1
+		flinkTableEnv = StreamTableEnvironment.create(env);
+	}
+
+	@Test
+	public void testReadFromQueryInternal() {
+		config.setFilter("word_count > 500");
+		config.setProjectId("q-gcp-6750-pso-gs-flink-22-01");
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java:53]: Reading from env variable

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+	public FlinkBigQueryIntegrationTestBase() {
+
+		this.bq = BigQueryOptions.getDefaultInstance().getService();
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		this.flinkTableEnv = StreamTableEnvironment.create(env);
+		this.config = new Configuration();
+		// TODO: change accordingly
+		this.config.setGcpCredentialKeyFile("C:\\sridhar\\GS\\q-gcp-6750-pso-gs-flink-22-01-1231782c49d3.json");
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java:38]: Reading "GOOGLE_APPLICATION_CREDENTIALS","GOOGLE_CLOUD_PROJECT","GOOGLE_CLOUD_DATASET", "GOOGLE_CLOUD_TABLE" from environment variables

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadByFormatIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		config.setSelectedFields("word_count,word");
+
+		String srcQueryString = "CREATE TABLE " + config.getBigQueryReadTable() + " (word STRING , word_count BIGINT)";
+		flinkTableEnv.executeSql(srcQueryString + "\n" + "WITH (\n" + "  'connector' = 'bigquery',\n"
+				+ "  'format' = 'arrow',\n" + "  'configOptions' = '" + config.getConfigMap() + "'\n" + ")");
+
+		Table result = flinkTableEnv.from(config.getBigQueryReadTable());
+
+		System.out.println(result.getSchema().getFieldDataType(1));
+		System.out.println(result.getResolvedSchema());
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadByFormatIntegrationTest.java:52]: Removed

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.junit.Test;
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.flink.bigquery.FlinkBigQueryException;
+import com.google.cloud.flink.bigquery.model.Configuration;
+
+public class FlinkReadFromQueryIntegrationTest extends FlinkBigQueryIntegrationTestBase {
+
+	private BigQuery bq;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadFromQueryIntegrationTest.java:35]: Addressed. Removed this class

File: src/main/java/com/google/cloud/flink/bigquery/arrow/util/ArrowToRowDataConverter.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.types.logical.ArrayType;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeUtils;
+
+@FunctionalInterface
+public interface ArrowToRowDataConverter extends Serializable {
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/arrow/util/ArrowToRowDataConverter.java:36]: Will address in the next PR

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+			this.allocator = new RootAllocator(Long.MAX_VALUE);
+		}
+		for (Field field : schema.getFields()) {
+			vectors.add(field.createVector(allocator));
+		}
+		root = new VectorSchemaRoot(vectors);
+		this.loader = new VectorLoader(root);
+	}
+
+	@Override
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:92]: Checking object is null to ensure it is end of stream and returns true.

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java (hunk 4 above)
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java:47]: renamed variable as per standard

File: src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+			log.error("Error while reading big query session", ex);
+			throw new FlinkBigQueryException("Error while reading big query session", ex);
+
+		}
+		return new BigQueryDynamicTableSource(decodingFormat, producedDataType);
+	}
+
+	@Override
+	public DynamicTableSink createDynamicTableSink(Context context) {
+		return null;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java:118]: Added TODO for write logic

File: src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+//			List<DataType> datatypeList = context.getCatalogTable().getResolvedSchema().getColumnDataTypes();
+//			List<String> columnNameList = context.getCatalogTable().getResolvedSchema().getColumnNames();
+//			if (datatypeList.indexOf(DataTypes.INT())>-1 ) {
+//				datatypeList.set(datatypeList.indexOf(DataTypes.INT()),DataTypes.BIGINT());
+//			}
+//			
+//			context.getCatalogTable().getResolvedSchema();
+			producedDataType = context.getCatalogTable().getResolvedSchema().toPhysicalRowDataType();
+			readSession = BigQueryReadSession.getReadsession(projectId, table, dataset, configOption);
+		} catch (IOException ex) {
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java:108]: Storing readSession value here and referencing the value from class "BigQuerySourceFunction.java", inside run() method

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java (hunk 5 above)
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java:29]: Instantiating the class in the base class and setting up few common configurations which are required in sub class implementations

File: src/main/java/com/google/cloud/flink/bigquery/arrow/util/ArrowArrayColumnVector.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	 * Container which is used to store the sequence of array values of a column to
+	 * read.
+	 */
+	private final ListVector listVector;
+
+	private final ColumnVector elementVector;
+
+	public ArrowArrayColumnVector(ListVector listVector, ColumnVector elementVector) {
+		this.listVector = Preconditions.checkNotNull(listVector, "ListVector is null");
+		this.elementVector = Preconditions.checkNotNull(elementVector, "ColumnVector is null");
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/arrow/util/ArrowArrayColumnVector.java:39]: Will take of it in the next PR. Currently removed this class.

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.flink.bigquery.model.Configuration;
+
+public class FlinkBigQueryIntegrationTestBase {
+
+	private BigQuery bq;
+	public static StreamTableEnvironment flinkTableEnv;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java:28]: Addressed. Removed as static and made it is protected.

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.flink.bigquery.model.Configuration;
+
+public class FlinkBigQueryIntegrationTestBase {
+
+	private BigQuery bq;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkBigQueryIntegrationTestBase.java:27]: made it is as protected

File: pom.xml [hunk 17]
```diff
@@ -379,7 +433,7 @@
                     </archive>
                 </configuration>
             </plugin>
-            <plugin>
+            <!--plugin>
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply pom.xml:436]: We encountered some build issues using this plugin. For time being, we commented this to build the application.

Below is the error message:

[INFO] --- spotless-maven-plugin:2.21.0:check (default) @ flink-bigquery-connector ---
[ERROR] Step 'google-java-format' found problem in 'src\main\java\com\google\cloud\flink\bigquery\FlinkBigQueryConfig.java':
You are running Spotless on JVM 8, which limits you to google-java-format 1.7.
If you upgrade your JVM to 11+, then you can use google-java-format 1.13.0, which may have fixed this problem.
java.lang.Exception: You are running Spotless on JVM 8, which limits you to google-java-format 1.7.
If you upgrade your JVM to 11+, then you can use google-java-format 1.13.0, which may have fixed this problem.
    at com.diffplug.spotless.Jvm$Support$1.apply (Jvm.java:181)

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.util.Preconditions;
+
+public class ArrowDeserializationSchema<T> implements DeserializationSchema<T>, Serializable {
+	private static final long serialVersionUID = 1L;
+
+	private BufferAllocator allocator;
+	private final TypeInformation<RowData> typeInfo;
+	static ArrowRecordBatch deserializedBatch;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:45]: Removed static modifier

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+		ArrowDeserializationSchema<?> that = (ArrowDeserializationSchema<?>) o;
+		return recordClazz.equals(that.recordClazz) && Objects.equals(schema, that.schema);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(recordClazz, schema);
+	}
+
+	public static void close() {
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:120]: Removed static modifier

File: src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+		if (message == null) {
+			throw new FlinkBigQueryException("Deserializing message is empty");
+		}
+		try {
+			VectorSchemaRoot root = nestedSchema.deserialize(message);
+			List<GenericRowData> rowdatalist = (List<GenericRowData>) runtimeConverter.convert(root);
+			for (int i = 0; i < rowdatalist.size(); i++) {
+				out.collect(rowdatalist.get(i));
+			}
+			ArrowDeserializationSchema.close();
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java:77]: Not referencing from other class. closing the root in the finally block.

File: src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java (hunk 2 above)
Comment (Sridhar-Durgam2022) [reply src/test/java/com/google/cloud/flink/bigquery/integration/FlinkReadIntegrationTest.java:36]: changed parameter  to tableResult

**Pull request closed without merging**